File: day6/day6.js
const fs = require('fs')

const args = process.argv.slice(2)

const input = fs.readFileSync(args[0]).toString().trim().split("\n")

input.forEach( line => 
  console.log(findMarker(line, 4))
)

function findMarker(input, markerLength) {
  let x = 0;
  let y = "";
  do {
    y = input.substring(x++, x + markerLength - 1);
  } while ( new Set(y).size != markerLength);

  return x + markerLength - 1;
}

File: day6/day6.nim
import strutils
import std/strbasics
import sequtils

proc findMarker(input: string, markerLength: int): int =
  var x: int = 0;
  var y: string = "";

  while deduplicate(toSeq(y.items)).len != markerLength:
    x = x + 1
    y = input.substr(x, x + markerLength - 1)

  return x + markerLength

var contents = readFile("day6.txt")

strip(contents)

let lines = contents.splitLines()

for line in lines:
  echo findMarker(line, 4)

